File: clients/go/protocol/logs.go
package protocol

import (
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strings"
)

// LogLevel is one rung of the severity ladder. The set is closed: it is the
// intersection of the ladders used by Go slog, Python logging, Rust tracing,
// pino and winston, so every bridge maps onto it without inventing a level.
type LogLevel string

// The severity ladder, least to most severe.
const (
	LevelTrace LogLevel = "trace"
	LevelDebug LogLevel = "debug"
	LevelInfo  LogLevel = "info"
	LevelWarn  LogLevel = "warn"
	LevelError LogLevel = "error"
	LevelFatal LogLevel = "fatal"
)

// LogLevels lists the ladder in order.
var LogLevels = []LogLevel{LevelTrace, LevelDebug, LevelInfo, LevelWarn, LevelError, LevelFatal}

// LogLevelSeverity is the numeric severity of each level; higher is worse.
var LogLevelSeverity = map[LogLevel]int{
	LevelTrace: 10, LevelDebug: 20, LevelInfo: 30,
	LevelWarn: 40, LevelError: 50, LevelFatal: 60,
}

// MaxLogAttrs bounds the number of attribute keys on one record.
const MaxLogAttrs = 64

var logLevelSet = map[string]struct{}{
	"trace": {}, "debug": {}, "info": {}, "warn": {}, "error": {}, "fatal": {},
}

var logRecordFields = []string{"ts", "level", "message", "attrs", "logger", "seq", "revision"}

// ValidLogLevel reports whether level is one of the six.
func ValidLogLevel(level LogLevel) bool {
	_, ok := logLevelSet[string(level)]
	return ok
}

// LogRecord is one application log record.
//
// TS is Unix epoch milliseconds, not session-relative: an adapter has no
// reliable view of when the driver considers the session to have started, so
// the wall clock is the only clock both sides agree on without negotiating.
// The driver rebases it onto the session timeline.
type LogRecord struct {
	TS      int64    `json:"ts"`
	Level   LogLevel `json:"level"`
	Message string   `json:"message"`
	// Attrs is flat structured context: scalars only, because nested values
	// make a record's size unbounded and depth-dependent. Bridges flatten.
	Attrs  map[string]any `json:"attrs,omitempty"`
	Logger string         `json:"logger,omitempty"`
	// Seq is a per-session counter assigned by the adapter. A gap tells the
	// driver records were dropped upstream, not lost in transit.
	Seq      int64 `json:"seq"`
	Revision int64 `json:"revision,omitempty"`
}

// LogMessage carries one record to the driver.
type LogMessage struct {
	Type   string     `json:"type"`
	Record *LogRecord `json:"record"`
}

// NewLogMessage wraps a record in its envelope.
func NewLogMessage(record *LogRecord) LogMessage {
	return LogMessage{Type: "log", Record: record}
}

// ValidateLogRecord checks an untrusted record against limits, mirroring
// ValidateSnapshot: measured against maxLogRecordBytes first, then checked
// field by field. Returns nil when the record is acceptable.
func ValidateLogRecord(value any, limits Limits) error {
	projected, err := ProjectDTO(value, limits.MaxDepth)
	if err != nil {
		code := "schema"
		if ViolationCode(err) == "dto-depth" {
			code = "depth"
		}
		return invalid(code, "%s", err.Error())
	}

	serialised, err := marshalCanonical(projected)
	if err != nil {
		return invalid("schema", "log record is not JSON-serialisable")
	}
	if len(serialised) > limits.MaxLogRecordBytes {
		return invalid("bytes", "log record is %d bytes, ceiling is %d", len(serialised), limits.MaxLogRecordBytes)
	}

	record, ok := projected.(map[string]any)
	if !ok {
		return invalid("schema", "log record must be an object")
	}

	for key := range record {
		known := false
		for _, field := range logRecordFields {
			if key == field {
				known = true
				break
			}
		}
		if !known {
			return invalid("schema", "unknown log record property %q", key)
		}
	}

	if ts, ok := safeNonNegative(record["ts"]); !ok || ts == 0 {
		return invalid("schema", "ts must be a positive safe integer (epoch milliseconds)")
	}
	level, _ := record["level"].(string)
	if _, known := logLevelSet[level]; !known {
		names := make([]string, 0, len(LogLevels))
		for _, item := range LogLevels {
			names = append(names, string(item))
		}
		return invalid("schema", "level must be one of %s", strings.Join(names, ", "))
	}
	message, ok := record["message"].(string)
	if !ok {
		return invalid("schema", "message must be a string")
	}
	if len(message) > limits.MaxStringBytes {
		return invalid("string-bytes", "message exceeds %d UTF-8 bytes", limits.MaxStringBytes)
	}
	if _, ok := safeNonNegative(record["seq"]); !ok {
		return invalid("schema", "seq must be a non-negative safe integer")
	}

	if logger, present := record["logger"]; present {
		text, ok := logger.(string)
		if !ok {
			return invalid("schema", "logger must be a string")
		}
		if len(text) > limits.MaxStringBytes {
			return invalid("string-bytes", "logger exceeds %d UTF-8 bytes", limits.MaxStringBytes)
		}
	}

	if revision, present := record["revision"]; present {
		if value, ok := safeNonNegative(revision); !ok || value == 0 {
			return invalid("revision", "revision must be a positive safe integer")
		}
	}

	if attrsValue, present := record["attrs"]; present {
		attrs, ok := attrsValue.(map[string]any)
		if !ok {
			return invalid("schema", "attrs must be a flat object")
		}
		if len(attrs) > MaxLogAttrs {
			return invalid("count", "attrs carries %d keys, ceiling is %d", len(attrs), MaxLogAttrs)
		}
		// Sorted so a rejection names the same key run to run.
		keys := make([]string, 0, len(attrs))
		for key := range attrs {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		for _, key := range keys {
			if len(key) > limits.MaxStringBytes {
				return invalid("string-bytes", "attribute key %q exceeds the string ceiling", key)
			}
			switch attr := attrs[key].(type) {
			case nil, bool:
			case string:
				if len(attr) > limits.MaxStringBytes {
					return invalid("string-bytes", "attribute %q exceeds the string ceiling", key)
				}
			case float64:
				if math.IsNaN(attr) || math.IsInf(attr, 0) {
					return invalid("schema", "attribute %q must be a finite number", key)
				}
			default:
				return invalid("schema", "attribute %q must be a string, number, boolean or null", key)
			}
		}
	}
	return nil
}

func safeNonNegative(value any) (float64, bool) {
	number, ok := value.(float64)
	if !ok || number != math.Trunc(number) || number < 0 || number > maxSafeInteger {
		return 0, false
	}
	return number, true
}

// Validate marshals the record and runs it through ValidateLogRecord, so an
// adapter checks exactly the bytes the driver will see.
func (r *LogRecord) Validate(limits Limits) error {
	body, err := marshalCanonical(r)
	if err != nil {
		return invalid("schema", "log record is not JSON-serialisable")
	}
	var parsed any
	if err := json.Unmarshal(body, &parsed); err != nil {
		return invalid("schema", "log record is not JSON-serialisable")
	}
	return ValidateLogRecord(parsed, limits)
}

// FlattenAttrs turns nested context into dotted keys, as the wire requires:
// {"db": {"host": "x"}} becomes {"db.host": "x"}. Values that are still not
// scalars afterwards are rendered as text, because losing a value's shape
// beats dropping the record that carries it.
func FlattenAttrs(value map[string]any) map[string]any {
	flat := make(map[string]any, len(value))
	flattenInto(flat, "", value, 0)
	return flat
}

func flattenInto(into map[string]any, prefix string, value map[string]any, depth int) {
	for key, item := range value {
		name := prefix + key
		switch typed := item.(type) {
		case map[string]any:
			if depth < 4 {
				flattenInto(into, name+".", typed, depth+1)
				continue
			}
			into[name] = fmt.Sprint(typed)
		case nil, bool, string:
			into[name] = typed
		case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
			into[name] = typed
		case float32, float64:
			into[name] = typed
		default:
			into[name] = fmt.Sprint(typed)
		}
	}
}

File: clients/rust/src/logs.rs
//! Application log records carried over the semantic channel.
//!
//! A TUI cannot print diagnostics to the screen without corrupting the render,
//! so applications write them to a logger instead. The `logs` capability
//! forwards those records to the driver, where they become assertable test
//! state rather than invisible side effects.
//!
//! Records are bounded exactly like snapshots: measured against a byte ceiling
//! and rejected wholesale on any violation, so a misbehaving logger degrades
//! into dropped records rather than unbounded driver memory.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::error::ValidationError;
use crate::framing::project_dto;
use crate::limits::Limits;
use crate::marker::MAX_SAFE_INTEGER;

/// One rung of the severity ladder.
///
/// The set is closed: it is the intersection of the ladders used by Rust
/// `tracing`, Python `logging`, Go `slog`, pino and winston, so every bridge
/// maps onto it without inventing a level.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LogLevel {
    /// Finer than debug.
    Trace,
    /// Diagnostic detail.
    Debug,
    /// Ordinary progress.
    Info,
    /// Something recoverable went wrong.
    Warn,
    /// An operation failed.
    Error,
    /// The application cannot continue.
    Fatal,
}

impl LogLevel {
    /// Numeric severity; higher is more severe.
    pub fn severity(self) -> u8 {
        match self {
            LogLevel::Trace => 10,
            LogLevel::Debug => 20,
            LogLevel::Info => 30,
            LogLevel::Warn => 40,
            LogLevel::Error => 50,
            LogLevel::Fatal => 60,
        }
    }

    /// The wire spelling.
    pub fn as_str(self) -> &'static str {
        match self {
            LogLevel::Trace => "trace",
            LogLevel::Debug => "debug",
            LogLevel::Info => "info",
            LogLevel::Warn => "warn",
            LogLevel::Error => "error",
            LogLevel::Fatal => "fatal",
        }
    }
}

/// The ladder in order, least to most severe.
pub const LOG_LEVELS: [&str; 6] = ["trace", "debug", "info", "warn", "error", "fatal"];

/// Maximum number of attribute keys on one record.
pub const MAX_LOG_ATTRS: usize = 64;

const RECORD_FIELDS: [&str; 7] = [
    "ts", "level", "message", "attrs", "logger", "seq", "revision",
];

/// A structured attribute value. Scalars only: nested values make a record's
/// size unbounded and depth-dependent, so bridges flatten before they send.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AttrValue {
    /// Absent or explicitly empty.
    Null,
    /// A flag.
    Bool(bool),
    /// An integer.
    Int(i64),
    /// A finite float.
    Float(f64),
    /// Text.
    Text(String),
}

impl From<bool> for AttrValue {
    fn from(value: bool) -> Self {
        AttrValue::Bool(value)
    }
}

impl From<i64> for AttrValue {
    fn from(value: i64) -> Self {
        AttrValue::Int(value)
    }
}

impl From<u64> for AttrValue {
    fn from(value: u64) -> Self {
        AttrValue::Int(value as i64)
    }
}

impl From<f64> for AttrValue {
    fn from(value: f64) -> Self {
        if value.is_finite() {
            AttrValue::Float(value)
        } else {
            AttrValue::Text(value.to_string())
        }
    }
}

impl From<&str> for AttrValue {
    fn from(value: &str) -> Self {
        AttrValue::Text(value.to_owned())
    }
}

impl From<String> for AttrValue {
    fn from(value: String) -> Self {
        AttrValue::Text(value)
    }
}

/// One application log record.
///
/// `ts` is Unix epoch milliseconds, not session-relative: an adapter has no
/// reliable view of when the driver considers the session to have started, so
/// the wall clock is the only clock both sides agree on without negotiating.
/// The driver rebases it onto the session timeline.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LogRecord {
    /// Unix epoch milliseconds when the record was produced.
    pub ts: i64,
    /// Severity.
    pub level: LogLevel,
    /// Human-readable message, already formatted by the source logger.
    pub message: String,
    /// Flat structured context; sorted so a record serialises the same way twice.
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub attrs: BTreeMap<String, AttrValue>,
    /// Logger or channel name, e.g. `http` or `db.pool`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub logger: Option<String>,
    /// Per-session counter assigned by the adapter. A gap tells the driver
    /// records were dropped upstream rather than lost in transit.
    pub seq: i64,
    /// Semantic revision current when the record was produced, when known.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub revision: Option<i64>,
}

impl LogRecord {
    /// A record with only the required fields. `seq` is assigned by the client.
    pub fn new(level: LogLevel, message: impl Into<String>) -> Self {
        Self {
            ts: 0,
            level,
            message: message.into(),
            attrs: BTreeMap::new(),
            logger: None,
            seq: 0,
            revision: None,
        }
    }

    /// Add one flat attribute.
    pub fn with_attr(mut self, key: impl Into<String>, value: impl Into<AttrValue>) -> Self {
        self.attrs.insert(key.into(), value.into());
        self
    }

    /// Name the logger this record came from.
    pub fn with_logger(mut self, logger: impl Into<String>) -> Self {
        self.logger = Some(logger.into());
        self
    }

    /// Validate against `limits`, as the driver will.
    ///
    /// # Errors
    /// Returns a [`ValidationError`] with the shared taxonomy.
    pub fn validate(&self, limits: &Limits) -> Result<(), ValidationError> {
        let value = serde_json::to_value(self)
            .map_err(|_| ValidationError::new("schema", "log record is not JSON-serialisable"))?;
        validate_log_record(&value, limits)
    }
}

fn fail(code: &'static str, detail: impl Into<String>) -> ValidationError {
    ValidationError::new(code, detail)
}

fn safe_non_negative(value: Option<&Value>) -> Option<i64> {
    value
        .and_then(Value::as_i64)
        .filter(|number| *number >= 0 && *number <= MAX_SAFE_INTEGER)
}

/// Validate an untrusted log record against `limits`.
///
/// Mirrors [`crate::validate_snapshot`]: projected first, then measured
/// against `max_log_record_bytes`, then checked field by field. Never panics.
///
/// # Errors
/// Returns a [`ValidationError`] whose `code` matches the reference
/// implementation's.
pub fn validate_log_record(value: &Value, limits: &Limits) -> Result<(), ValidationError> {
    if let Err(violation) = project_dto(value, limits.max_depth) {
        let code = if violation.code == "dto-depth" {
            "depth"
        } else {
            "schema"
        };
        return Err(fail(code, violation.to_string()));
    }

    let serialised = serde_json::to_vec(value)
        .map_err(|_| fail("schema", "log record is not JSON-serialisable"))?;
    if serialised.len() > limits.max_log_record_bytes {
        return Err(fail(
            "bytes",
            format!(
                "log record is {} bytes, ceiling is {}",
                serialised.len(),
                limits.max_log_record_bytes
            ),
        ));
    }

    let Some(record) = value.as_object() else {
        return Err(fail("schema", "log record must be an object"));
    };

    for key in record.keys() {
        if !RECORD_FIELDS.contains(&key.as_str()) {
            return Err(fail(
                "schema",
                format!("unknown log record property \"{key}\""),
            ));
        }
    }

    match safe_non_negative(record.get("ts")) {
        Some(ts) if ts > 0 => {}
        _ => {
            return Err(fail(
                "schema",
                "ts must be a positive safe integer (epoch milliseconds)",
            ))
        }
    }
    match record.get("level").and_then(Value::as_str) {
        Some(level) if LOG_LEVELS.contains(&level) => {}
        _ => {
            return Err(fail(
                "schema",
                format!("level must be one of {}", LOG_LEVELS.join(", ")),
            ))
        }
    }
    let Some(message) = record.get("message").and_then(Value::as_str) else {
        return Err(fail("schema", "message must be a string"));
    };
    if message.len() > limits.max_string_bytes {
        return Err(fail(
            "string-bytes",
            format!("message exceeds {} UTF-8 bytes", limits.max_string_bytes),
        ));
    }
    if safe_non_negative(record.get("seq")).is_none() {
        return Err(fail("schema", "seq must be a non-negative safe integer"));
    }

    if let Some(logger) = record.get("logger") {
        let Some(text) = logger.as_str() else {
            return Err(fail("schema", "logger must be a string"));
        };
        if text.len() > limits.max_string_bytes {
            return Err(fail(
                "string-bytes",
                format!("logger exceeds {} UTF-8 bytes", limits.max_string_bytes),
            ));
        }
    }

    if let Some(revision) = record.get("revision") {
        match safe_non_negative(Some(revision)) {
            Some(value) if value > 0 => {}
            _ => return Err(fail("revision", "revision must be a positive safe integer")),
        }
    }

    if let Some(attrs) = record.get("attrs") {
        let Some(attrs) = attrs.as_object() else {
            return Err(fail("schema", "attrs must be a flat object"));
        };
        if attrs.len() > MAX_LOG_ATTRS {
            return Err(fail(
                "count",
                format!(
                    "attrs carries {} keys, ceiling is {MAX_LOG_ATTRS}",
                    attrs.len()
                ),
            ));
        }
        for (key, attr) in attrs {
            if key.len() > limits.max_string_bytes {
                return Err(fail(
                    "string-bytes",
                    format!("attribute key \"{key}\" exceeds the string ceiling"),
                ));
            }
            match attr {
                Value::Null | Value::Bool(_) => {}
                Value::Number(number) => {
                    if number.as_f64().is_none_or(|value| !value.is_finite()) {
                        return Err(fail(
                            "schema",
                            format!("attribute \"{key}\" must be a finite number"),
                        ));
                    }
                }
                Value::String(text) => {
                    if text.len() > limits.max_string_bytes {
                        return Err(fail(
                            "string-bytes",
                            format!("attribute \"{key}\" exceeds the string ceiling"),
                        ));
                    }
                }
                _ => {
                    return Err(fail(
                        "schema",
                        format!("attribute \"{key}\" must be a string, number, boolean or null"),
                    ))
                }
            }
        }
    }

    Ok(())
}
